File: native/android/MainActivity.java
package io.customli.sbbpos;

import android.Manifest;
import android.content.pm.PackageManager;
import android.os.Build;
import android.os.Bundle;

import com.getcapacitor.BridgeActivity;

import java.util.ArrayList;
import java.util.List;

public class MainActivity extends BridgeActivity {
  private static final int BLUETOOTH_PERMISSION_REQUEST = 1001;

  @Override
  public void onCreate(Bundle savedInstanceState) {
    registerPlugin(ThermalPrinterPlugin.class);
    super.onCreate(savedInstanceState);
    requestBluetoothPermissionsIfNeeded();
  }

  private void requestBluetoothPermissionsIfNeeded() {
    if (Build.VERSION.SDK_INT < Build.VERSION_CODES.S) return;

    List<String> missing = new ArrayList<>();
    if (checkSelfPermission(Manifest.permission.BLUETOOTH_CONNECT) != PackageManager.PERMISSION_GRANTED) {
      missing.add(Manifest.permission.BLUETOOTH_CONNECT);
    }
    if (checkSelfPermission(Manifest.permission.BLUETOOTH_SCAN) != PackageManager.PERMISSION_GRANTED) {
      missing.add(Manifest.permission.BLUETOOTH_SCAN);
    }

    if (!missing.isEmpty()) {
      requestPermissions(missing.toArray(new String[0]), BLUETOOTH_PERMISSION_REQUEST);
    }
  }
}

File: native/android/MainActivity.kt
package io.customli.sbbpos

import android.Manifest
import android.content.pm.PackageManager
import android.os.Build
import android.os.Bundle
import com.getcapacitor.BridgeActivity

class MainActivity : BridgeActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        registerPlugin(ThermalPrinterPlugin::class.java)
        super.onCreate(savedInstanceState)

        // Android 12+ splits classic Bluetooth access into the Nearby Devices
        // runtime permission group. The printer bridge needs both permissions:
        // CONNECT for bonded-device details/RFCOMM and SCAN for cancelling any
        // discovery before opening the socket. Request the complete set here so
        // the first visit to Printer Settings is immediately usable.
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            val missing = mutableListOf<String>()
            if (checkSelfPermission(Manifest.permission.BLUETOOTH_CONNECT) != PackageManager.PERMISSION_GRANTED) {
                missing.add(Manifest.permission.BLUETOOTH_CONNECT)
            }
            if (checkSelfPermission(Manifest.permission.BLUETOOTH_SCAN) != PackageManager.PERMISSION_GRANTED) {
                missing.add(Manifest.permission.BLUETOOTH_SCAN)
            }
            if (missing.isNotEmpty()) {
                requestPermissions(missing.toTypedArray(), 1001)
            }
        }
    }
}
